### PR TITLE
flex_trits_insert: check for aligned trits to use memcpy

### PR DESCRIPTION
**Pull request by semenov-vladyslav**
Added a special case in `flex_trits_insert` and `flex_trits_insert_from_pos` such that `memcpy` is used when source and destination flex trits are aligned.

# Test Plan:
Added tests (`test_flex_trits_insert`) which cover boundary conditions.
Also, need to run benchmarks and profiling.

### CODE REVIEW

File: common/trinary/flex_trit.c [hunk 1]
```diff
@@ -98,8 +98,17 @@ size_t flex_trits_insert(flex_trit_t *const to_flex_trits, size_t const to_len,
 #if defined(FLEX_TRIT_ENCODING_1_TRIT_PER_BYTE)
   memcpy(to_flex_trits + start, flex_trits, num_trits);
 #else
-  for (size_t i = 0; i < num_trits; i++) {
-    flex_trits_set_at(to_flex_trits, to_len, start + i, flex_trits_at(flex_trits, len, i));
+  if (0 == (start % NUM_TRITS_PER_FLEX_TRIT)) {
```
Comment (tsvisabo) [inline common/trinary/flex_trit.c:101]: Can we not always "handle head" so we get "new" start index and always memcpy from there?

**Review by tsvisabo — CHANGES_REQUESTED** [review]

File: common/trinary/flex_trit.c (hunk 1 above)
Comment (tsvisabo) [reply common/trinary/flex_trit.c:101]: actually no :(

**Review by tsvisabo — APPROVED** [review]

**Comment by thibault-martinez** [comment]
Before we merge I'd be interested in a benchmark, even very basic just
timing the test you wrote with and without your change.

Le mar. 23 juil. 2019 à 08:48, Tsvi Sabo <notifications@github.com> a
écrit :

**Comment by semenov-vladyslav** [comment]
`flex_trits_insert` is the bottleneck in `transaction_serialize`. All the chunks of a transaction are of fixed size and are multiple of 3, which make it perfect for TE3: all the buffers are aligned and `memcpy` is used. It's easy to confirm: 
- 1M serializations of a transaction *without* optimization takes approx 90 seconds, 
- *with* optimization -- a bit more than 1 second.

**Review by thibault-martinez — APPROVED** [review]
